File: service/src/main/java/uk/ac/bbsrc/earlham/browser/service/ajax/BlastGrassroot.java
/*
#
# Copyright (c) 2013. The Genome Analysis Centre, Norwich, UK
# TGAC Browser project contacts: Anil Thanki, Xingdong Bian, Robert Davey @ Earlham Institute
# **********************************************************************
#
# This file is part of TGAC Browser.
#
# TGAC Browser is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# TGAC Browser is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with TGAC Browser.  If not, see <http://www.gnu.org/licenses/>.
#
# ***********************************************************************
#
 */
package uk.ac.bbsrc.earlham.browser.service.ajax;

import net.sf.json.JSONArray;
import net.sf.json.JSONObject;
import net.sourceforge.fluxion.ajax.Ajaxified;
import net.sourceforge.fluxion.ajax.util.JSONUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.http.HttpSession;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Collections;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created with IntelliJ IDEA.
 * User: thankia
 * Date: 5/8/13
 * Time: 12:25 PM
 * To change this template use File | Settings | File Templates.
 */

@Ajaxified

public class BlastGrassroot {

    private Logger log = LoggerFactory.getLogger(getClass());

    /**
     * Method to get Params from Grassroot backend
     * At the moment not in use
     * @param session
     * @param json
     * @return
     * @throws IOException
     */
    public JSONObject getParams(HttpSession session, JSONObject json) throws IOException {

        JSONObject query = new JSONObject();

        JSONObject response = new JSONObject();
        JSONObject operation = new JSONObject();
        JSONArray services = new JSONArray();
        JSONObject service = new JSONObject();

        operation.put("operation", "get_named_service");
        service.put("so:alternateName","blast-blastn");
        services.add(service);

        query.put("operations", operation);
        query.put("services", services);
        response.put("query", query);

        String urlParameters = query.toString();

        URL url = new URL("https://grassroots.tools/dev/public_backend");

        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setDoOutput(true);
        connection.setDoInput(true);
        connection.setInstanceFollowRedirects(false);
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
        connection.setRequestProperty("charset", "utf-8");
        connection.setRequestProperty("Content-Length", "" + Integer.toString(urlParameters.getBytes().length));
        connection.setUseCaches(false);
        DataOutputStream wr = new DataOutputStream(connection.getOutputStream());
        wr.writeBytes(urlParameters);
        wr.flush();
        wr.close();
        BufferedReader input = new BufferedReader(new InputStreamReader(connection.getInputStream()));

        String result = "";
        String str;

        while (null != (str = input.readLine())) {
            result += str;
        }
        input.close();
        connection.disconnect();
        response.put("response", result);

        JSONObject result_obj = response.getJSONObject("response");

        JSONArray groups = result_obj.getJSONArray("services").getJSONObject(0).getJSONObject("operation").getJSONObject("parameter_set").getJSONArray("groups");

        JSONArray parameters = result_obj.getJSONArray("services").getJSONObject(0).getJSONObject("operation").getJSONObject("parameter_set").getJSONArray("parameters");

        response.put("parameters", parameters);
        JSONObject params = new JSONObject();
        for (int i=0; i< parameters.size(); i++){
            JSONObject param = parameters.getJSONObject(i);
            String key = param.getString("group");
            if(params.containsKey(key)){
                params.getJSONArray(key).addAll(Collections.singleton(param));
            }else{
                params.put(key,Collections.singleton(param));
            }
        }

        response.put("groups", groups);
        response.put("params", params);

        return response;
    }

    /**
     * Return JSONObject
     * <p>
     * Generate url for Grassroot BLAST submission and
     * make a HTTPCOnnection to submit a job to Grassroot BLAST
     * </p>
     *
     * @param session an HTTPSession comes from ajax call
     * @param json    json object with key parameters sent from ajax call
     * @return JSONObject with NCBI submission id
     * @throws IOException
     */
    public JSONObject BlastSearchSequence(HttpSession session, JSONObject json) throws IOException {
        try {
            String blastdb = json.getString("blastdb");
            StringBuilder sb = new StringBuilder();
            String query = json.getString("querystring");
            String type = json.getString("type");
            String BlastAccession = json.getString("BlastAccession");
            JSONObject in_params = json.getJSONObject("params");
            int format = json.getInt("format");
            query = query.replaceAll(">+", "#>");

            JSONArray services = new JSONArray();
            JSONObject service = new JSONObject();

            service.put("start_service", true);
            service.put("so:alternateName", "blast-blastn");

            JSONObject parameter_set = new JSONObject();
            JSONArray parameters = new JSONArray();

            JSONObject param = new JSONObject();

            param.put("param", "job_id");
            param.put("current_value", null);

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "query");
            param.put("current_value", query);


            parameters.add(param);
            param = new JSONObject();
            param.put("param", "subrange_from");
            param.put("current_value", null);

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "subrange_to");
            param.put("current_value", null);

            parameters.add(param);

            param = new JSONObject();
            param.put("param", "max_target_seqs");
            param.put("current_value", 5);

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "evalue");
            param.put("current_value", 10);

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "outfmt");
            param.put("current_value", format);

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "word_size");
            param.put("current_value", in_params.getInt("word_size"));

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "task");
            param.put("current_value", type);

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "reward");
            param.put("current_value", in_params.getInt("reward"));

            parameters.add(param);
            param = new JSONObject();
            param.put("param", "penalty");
            param.put("current_value", in_params.getInt("penalty"));

            parameters.add(param);

            param = new JSONObject();
            param.put("param", "Available Databases provided by EI Grassroots development server -> IWGSC v1.0");

            param.put("current_value", true);

            parameters.add(param);

            parameter_set.put("level", "simple");
            parameter_set.put("parameters", parameters);


            service.put("parameter_set", parameter_set);


            services.add(service);

            JSONObject params = new JSONObject();
            params.put("services", services);

            String urlParameters = params.toString();

            JSONObject result_obj = new JSONObject();

            log.info("\n\n\n\n\t\t\t urlParameters" + urlParameters);
            result_obj.put("urlParameters", urlParameters);


            URL url = new URL("https://grassroots.tools/dev/public_backend");

            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setDoOutput(true);
            connection.setDoInput(true);
            connection.setInstanceFollowRedirects(false);
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
            connection.setRequestProperty("charset", "utf-8");
            connection.setRequestProperty("Content-Length", "" + Integer.toString(urlParameters.getBytes().length));
            connection.setUseCaches(false);
            DataOutputStream wr = new DataOutputStream(connection.getOutputStream());
            wr.writeBytes(urlParameters);
            wr.flush();
            wr.close();
            BufferedReader input = new BufferedReader(new InputStreamReader(connection.getInputStream()));

            String result = "";
            String str;

            while (null != (str = input.readLine())) {
                result += str;
            }
            input.close();
            connection.disconnect();


            result_obj.put("result_string", result);
            JSONObject web = result_obj.getJSONObject("result_string");
            result_obj.put("result", result);//.getJSONArray("results").getJSONObject(0).getString("job_uuid"));
            result_obj.put("id", web.getJSONArray("results").getJSONObject(0).getString("job_uuid"));
            result_obj.put("BlastAccession", BlastAccession);

            return result_obj;
        } catch (Exception e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
            return JSONUtils.SimpleJSONError(e.getMessage());
        }

    }

    /**
     * Return JSONObject
     * <p>
     * call connectGrassRoot to check BLAST is finished or running
     * if finished call method parseNCBI
     * </p>
     *
     * @param session an HTTPSession comes from ajax call
     * @param json    json object with key parameters sent from ajax call
     * @return JSONObject with BLAST output converted into tabular format
     * @throws IOException
     */
    public JSONObject BlastGetResult(HttpSession session, JSONObject json) throws IOException {
        try {

            String BlastAccession = json.getString("BlastAccession");
            String id = json.getString("id");

            JSONObject url_req = new JSONObject();
            JSONObject operation = new JSONObject();
            String[] ids = {id};
            operation.put("operation", 5);
            url_req.put("operations", operation);
            url_req.put("services", ids);

            JSONObject jsonObject = new JSONObject();
            String urlParameters = url_req.toString();

            String str;
            str = connectGrassRoot(urlParameters);

            jsonObject.put("id", id);
            jsonObject.put("response", str);
            jsonObject.put("BlastAccession", BlastAccession);

            return jsonObject;
        } catch (Exception e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
            return JSONUtils.SimpleJSONError(e.getMessage());
        }


    }

    /**
     * Return String
     * <p>
     * Connect to Grassroot for BLAST result and decide from RegEx that BLAST finished or running
     * </p>
     *
     * @param urlParameters String urlParameters for NCBI connection
     * @return String running/finished
     */
    private String connectGrassRoot(String urlParameters) {

        try {

            log.info("\n\n\n\n\t connectNCBI "+ urlParameters);
            URL url = new URL("https://grassroots.tools/dev/public_backend");

            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setDoOutput(true);
            connection.setDoInput(true);
            connection.setInstanceFollowRedirects(false);
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
            connection.setRequestProperty("charset", "utf-8");
            connection.setRequestProperty("Content-Length", "" + Integer.toString(urlParameters.getBytes().length));
            connection.setUseCaches(false);

            DataOutputStream wr = new DataOutputStream(connection.getOutputStream());
            wr.writeBytes(urlParameters);
            wr.flush();
            wr.close();
            BufferedReader input = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            String str;
            String result = "";

            while (null != (str = input.readLine())) {
                result += str;
            }

            input.close();
            connection.disconnect();
            return result;
        } catch (Exception e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
            String sb = e.toString();
            return sb;
        }
    }

}

File: browser-web/src/main/webapp/scripts/blast_grassroot.js
/*
*
* Copyright (c) 2013. Earlham Institute, Norwich, UK
* TGAC Browser project contacts: Anil Thanki, Xingdong Bian, Robert Davey @ Earlham Institute
* **********************************************************************
*
* This file is part of TGAC Browser.
*
* TGAC Browser is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* TGAC Browser is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with TGAC Browser.  If not, see <http://www.gnu.org/licenses/>.
*
* ***********************************************************************
*
*/
/**
 * Created with IntelliJ IDEA.
 * User: thankia
 * Date: 5/8/13
 * Time: 12:29 PM
 * To change this template use File | Settings | File Templates.
 */


function getParams(type) {
    ajaxurl = '/' + jQuery('#title').text() + '/' + jQuery('#title').text() + '/fluxion.ajax';
    Fluxion.doAjax(
        'blastGrassroot',
        'getParams',
        {
            'url': ajaxurl,
            'type': type
        },
        {
            'doOnSuccess': function (json) {
                console.log(json)
            },
            'doOnError': function (json) {
                alert(json.error);
            }
        });
}


function blastSearch(query, db, type, params) {
    ajaxurl = '/' + jQuery('#title').text() + '/' + jQuery('#title').text() + '/fluxion.ajax';
    var link = "";
    var id = randomString(8);
    var format = 7;// qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sseq qseq";//"format";
    jQuery("#blast_list").append("<div id='" + id + "' class='blast_list_node'> <b>BLAST job " + id + " </b> is running. <p> <center><img style='position: relative;' src='./images/browser/loading_big.gif' height=15px alt='Loading'></center></div>")

    Fluxion.doAjax(
        'blastGrassroot',
        'BlastSearchSequence',
        {
            'url': ajaxurl,
            'querystring': query,
            'blastdb': db,
            'location': link,
            'BlastAccession': id,
            'format': format,
            'type': type,
            'params': params
        },
        {
            'doOnSuccess': function (json) {
                grassrootBLASTResult(json.BlastAccession, json.id)
            },
            'doOnError': function (json) {
                alert(json.error);
                jQuery("#" + id).html("<b>BLAST job " + id + "</b><br> Failed.> </span> ")
            }
        });
}


function grassrootBLASTResult(BlastAccession, id) {
    ajaxurl = '/' + jQuery('#title').text() + '/' + jQuery('#title').text() + '/fluxion.ajax';

    var format = "format";
    Fluxion.doAjax(
        'blastGrassroot',
        'BlastGetResult',
        {'url': ajaxurl, 'BlastAccession': BlastAccession, 'id': id, 'format': format},
        {
            'doOnSuccess': function (json) {
                jQuery("#notifier").hide()
                jQuery("#notifier").html("");

                if (json.response[0]["status_text"] == "Succeeded") {
                    jQuery("#" + json.BlastAccession).html("BLAST job " + json.BlastAccession + " <span title=\"Finished\" class=\"ui-button ui-icon ui-icon-check\"></span> <br>  <span onclick=toogleTable('" + json.BlastAccession + "') class=\"ui-button ui-icon ui-icon-zoomin\" > </span> <span onclick=deleteTable('" + json.BlastAccession + "') class=\"ui-button ui-icon ui-icon-trash\" > </span> ")
                    jQuery('#main').animate({"height": "0px"}, {duration: 300, queue: false});
                    jQuery('#main').fadeOut();
                    if (json.response[0]["results"][0]["data"].indexOf("0 hits found") > 0) {
                        jQuery("#" + json.BlastAccession).html("<b>BLAST job " + json.BlastAccession + "</b><br> No hits found. <span style=\"float: right; position: relative;\" onclick=deleteTable('" + json.id + "') class=\"ui-button ui-icon ui-icon-trash\" > </span> ")
                    } else {
                        parseGrassRootBLAST(json.response[0]["results"][0]["data"], json.BlastAccession);
                    }
                } else if (json.response[0]["status_text"] == "Started" || json.response[0]["status_text"] == "Pending") {

                    setTimeout(function () {
                        grassrootBLASTResult(BlastAccession, id)
                    }, 5000);
                } else {
                    var msg = json.response[0]["errors"]["runtime_errors"]["errors"][0];

                    jQuery("#" + json.BlastAccession).html("<b>BLAST job " + json.BlastAccession + "</b><br> Failed")

                    var errmsg = document.createElement('span');
                    errmsg.className = "ui-button ui-icon ui-icon-info";
                    errmsg.onclick = (function () {
                        return function () {
                            alert(msg);
                        }
                    })();
                    jQuery("#" + json.BlastAccession).append(errmsg)

                    var delEntry = document.createElement('span');
                    delEntry.className = "ui-button ui-icon ui-icon-trash";
                    delEntry.onclick = (function () {
                        return function () {
                            deleteTable(json.BlastAccession)
                        }
                    })();
                    jQuery("#" + json.BlastAccession).append(delEntry)
                }
            },
            'doOnError': function (json) {
                alert(json.error);
            }
        });
}

function parseGrassRootBLAST(result, id) {
    var hits = result.split("\n")

    if (hits.length > 1) {
        jQuery('#blastresult').fadeIn();
        jQuery('#blastresult').append("<table style=\"display: none;\" class='list' id='blasttable" + id + "'> <thead></thead><tbody></tbody></table>")
        //      "<tr><th> Query id </th> <th> Subject id </th>  <th> alignment length </th>  <th> mismatches </th>  <th> gap openings </th>  <th> q.start </th>  <th> q.end </th>  <th> s.start </th>  <th> s.end </th> <th> e-value </th> <th> bit score </th> <th> Subject db </th><th> Download Sequence </th>        </tr>        </thead>        <tbody>        </tbody>    </table>")

        for (var i = 0; i < hits.length; i++) {
            if (hits[i].indexOf("#") == 0) {
                if (hits[i].indexOf("Fields") > 0) {
                    var row = hits[i].replace("# Fields: ", "");
                    jQuery("#blasttable" + id + " thead").append("<tr><th>" + row.replaceAll(",", "<th>"))
                }
            } else {
                var cols = hits[i].split("\t");
                var col_html = "<tr>"
                for (var col = 0; col < cols.length; col++) {
                    if (col == 1) {
                        var from = parseInt(cols[8]) - 100;
                        var to = parseInt(cols[9]) + 100;
                        col_html += "<td><a target='_blank' href='../tgac_browser/index.jsp?query=" + cols[col] + "&&from=" + from + "&&to=" + to + "'>" + cols[col] + "</a></td>";
                    } else {
                        col_html += "<td>" + cols[col] + "</td>";
                    }
                }
                jQuery("#blasttable" + id + " tbody").append(col_html)
            }
        }
    } else {
        jQuery("#" + id).html("<b>BLAST job " + json.id + "</b><br> No hits found. <span style=\"float: right; position: relative;\" onclick=deleteTable('" + json.id + "') class=\"ui-button ui-icon ui-icon-trash\" > </span> ")
    }

    jQuery("#blasttable" + json.id).tablesorter();
    jQuery("'#blasttable" + json.id + "'").trigger("update");
}


function blastTrackSearch(query, start, end) {
    jQuery("#notifier").html("<img src='images/browser/loading2.gif' height='10px'> BLAST running ");
    jQuery("#notifier").show();
    ajaxurl = '/' + jQuery('#title').text() + '/' + jQuery('#title').text() + '/fluxion.ajax';
    var link = path;
    var id = randomString(8);
    var format = 7;//"format";
    if (!window['blasttrack']) {

        setBLASTTrack()

    }

    Fluxion.doAjax(
        'blastGrassroot',
        'BlastSearchSequence',
        {
            'url': ajaxurl,
            'querystring': query,
            'blastdb': 'db',
            'type': 'blastn',
            'location': link,
            'BlastAccession': id,
            'format': format,
            'custom_outfmt': 'qseqid sseqid qstart qend sstart send bitscore qseq sseq btop'
        },
        {
            'doOnSuccess': function (json) {
                // blastsdata.push(
                //     {
                //         id: json.html,
                //         start: start,
                //         end: end,
                //         db: db,
                //         hit: hit,
                //         link: link,
                //         format: 5,
                //         type: type
                //     });
                console.log("hererere")
                grassrootBLASTResultTrack(json.BlastAccession, json.id, start)

            },
            'doOnError': function (json) {
                alert(json.error);
            }
        });
}

function grassrootBLASTResultTrack(BlastAccession, id, start) {
    console.log("grassrootBLASTResultTrack")
    ajaxurl = '/' + jQuery('#title').text() + '/' + jQuery('#title').text() + '/fluxion.ajax';

    var format = "format";
    Fluxion.doAjax(
        'blastGrassroot',
        'BlastGetResult',
        {'url': ajaxurl, 'BlastAccession': BlastAccession, 'id': id, 'format': format},
        {
            'doOnSuccess': function (json) {
                jQuery("#notifier").hide()
                jQuery("#notifier").html("");

                if (json.response[0]["status_text"] == "Succeeded") {
                    parseGrassRootBLASTTrack(json.response[0]["results"][0]["data"], json.BlastAccession, start);
                } else if (json.response[0]["status_text"] == "Started" || json.response[0]["status_text"] == "Pending") {

                    setTimeout(function () {
                        grassrootBLASTResultTrack(BlastAccession, id, start)
                    }, 5000);
                } else {
                    alert("else")
                }
            },
            'doOnError': function (json) {
                alert(json.error);
            }
        });
}


function parseGrassRootBLASTTrack(result, id, start) {
    var hits = result.split("\n")
    var blast = [];
    for (var i = 0; i < hits.length; i++) {
        if (hits[i].indexOf("#") == 0) {
            if (hits[i].indexOf("Fields") > 0) {
                var row = hits[i].replace("# Fields: ", "");
            }
        } else {
            var row = hits[i].split("\t");
            var each_track = {};
            each_track['start'] = parseInt(start) + parseInt(row[2]);
            each_track['end'] = parseInt(start) + parseInt(row[3]);
            each_track['desc'] = row[1] + ":" + row[4] + "-" + row[5];
            each_track['score'] = row[6];
            each_track['indels'] = []
            row[9] = row[9].replace(/(\d+)/g, " $1 ");
            row[9] = row[9].replace(" ", "")
            row[9] = row[9].substring(0, row[9].length - 1);
            var btop_array = row[9].split(" ");

            var indels = []
            var pos = 0;
            var reg = /^\d+$/;
            for (var j = 0; j < btop_array.length; j++) {
                var indel = {}
                if (reg.test(btop_array[j])) {
                    pos = Number(pos) + Number(btop_array[j])
                } else {
                    if (btop_array[j].length == 2) {
                        pos = Number(pos) + Number(btop_array[j].length / 2)
                        var from = pos - 4;
                        indel['position'] = pos
                        indel['query'] = row[7].substr(from, 7)
                        indel['hit'] = row[8].substr(from, 7)
                        indels.push(indel)
                    } else {
                        var spos = pos;
                        for (var k = 0; k < btop_array[j].length; k++) {
                            if (k % 2 == 0) {
                                //query
                                if (btop_array[j].substr(k, 1) == "-") {
                                } else {
                                    pos = Number(pos) + 1;
                                }
                            } else {
                                //subk=ject
                                if (btop_array[j].substr(k, 1) == "-") {
                                } else {
                                    spos = Number(spos) + 1;
                                }
                            }
                        }

                        indel['position'] = pos
                        var from = pos - 4;
                        var sfrom = spos - 4;
                        indel['query'] = row[7].substr(from, Number(btop_array[j].length) + Number(3))
                        indel['hit'] = row[8].substr(sfrom, Number(btop_array[j].length) + Number(3))
                        indels.push(indel)
                    }
                    each_track['indels'] = indels
                }
            }
            blast.push(each_track);
            if (blast.length > 20) {
                break;
            }

        }
    }
    if (!window['blasttrack']) {
        window['blasttrack'] = "running";
    }
    if (window['blasttrack'] == "running") {
        window['blasttrack'] = blast;
    } else {
        jQuery.merge(window['blasttrack'], blast);
    }
    jQuery('input[name=blasttrackCheckbox]').attr('checked', true);
    trackToggle("blasttrack");
}